### ISSUE TITLE
Code on SQL building path should not throw exceptions

### ISSUE THREAD

**Issue by knutwannheden** [jOOQ/jOOQ#8640]
There are some examples of code in package `org.jooq.impl` throwing exceptions in the code path building SQL (i.e. called inside the implementation of `QueryPartInternal#accept(Context)`). Two such examples are:

* `Expression#getIntervalExpression(Configuration)` which will throw a `SQLDialectNotSupportedException` when called for `INGRES`
* `FunctionTable#accept(Context)` also throws an `SQLDialectNotSupportedException` when called for unsupported dialects

There are also other cases where other exception types get thrown.

It would be better if these methods did not throw any exceptions. If the generated SQL doesn't execute that is OK. After all the `@Support` annotations already document what isn't supported.

**Comment by lukaseder** [comment]
I definitely agree with this. Other instances where generated SQL doesn't make sense is for example when an `UPDATE` statement doesn't contain any `SET` clause. `FieldMapForUpdate` then generates the following invalid SQL:

```java
ctx.sql("[ no fields are updated ]");
```

I think it's worth searching the code base for all exceptions being thrown from `accept()` methods, list them here, and fix them once. Without any expectation towards exhaustiveness, of course.

**Comment by knutwannheden** [comment]
> ```java
> ctx.sql("[ no fields are updated ]");
> ```

Yes, I suppose something like this should be possible in most cases rather than throwing an exception.

**Comment by knutwannheden** [comment]
Updated issue description with the places in the code where exceptions get thrown. There could of course be others I didn't find.

**Comment by lukaseder** [comment]
Quite a few places! I'm marking this as an incompatible change as users might have grown used to catching such exceptions, which will no longer occur - at least not in the same way

**Comment by lukaseder** [comment]
> Do we have an example of using `FunctionTable`?

`DSL::unnest0`

**Comment by knutwannheden** [comment]
> `DSL::unnest0`

I found this call site, but didn't find any integration test using this. Possibly that is just missing.

**Comment by lukaseder** [comment]
> I found this call site, but didn't find any integration test using this. Possibly that is just missing.

`testPostgresUnnestArrayOfRecord`

**Comment by lukaseder** [comment]
> I found this call site, but didn't find any integration test using this.

Look at `AbstractRoutine.RoutineField` usage

**Comment by knutwannheden** [comment]
Hmm, I still couldn't find any integration test, which actually ends up calling `FunctionTable#accept()`, but I now better understand the use case.

**Comment by lukaseder** [comment]
Looking into this now. Some of these exceptions are OK, because they should never be thrown, e.g.:

> `CommentOnImpl#acceptDefault()` (line 222)

The code being:

```java
if (table != null)
    ctx.visit(isView && SUPPORTS_COMMENT_ON_VIEW.contains(ctx.dialect()) ? K_VIEW : K_TABLE)
       .sql(' ')
       .visit(table);
else if (field != null)
    ctx.visit(K_COLUMN).sql(' ').visit(field);
else
    throw new IllegalStateException();
```

I.e. the `else` case is a bug in jOOQ, not in client code. We want to see this exception in integration tests, if there is every a case where both `table` and `field` are `null`

**Comment by lukaseder** [comment]
A related issue: https://github.com/jOOQ/jOOQ/issues/13541
